### PR TITLE
meta: refactor into a package

### LINKED ISSUE #1695: Refactor meta into package
Estimate: 2 days

Currently Snapcraft requires all its metadata (snap name, version, etc.) to be specified in the snapcraft.yaml. However, lots of parts already contain such information, whether from a setup.py, XML, AppStream, etc.

I’m considering this task to only be for existing metadata (name, version, summary, etc.). Adding support for an AppStream ID etc. is a separate task.

### PR DESCRIPTION
**Pull request by elopio**
- [x] Have you followed the [guidelines for contributing](https://github.com/snapcore/snapcraft/blob/master/CONTRIBUTING.md)?
- [x] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [x] If this is a bugfix. Have you checked that there is a bug report open for the issue you are trying to fix on [bug reports](https://bugs.launchpad.net/snapcraft)?
- [x] If this is a new feature. Have you discussed the design on the [forum](https://forum.snapcraft.io)?
- [x] Have you successfully run `./runtests.sh static`?
- [x] Have you successfully run `./runtests.sh unit`?

-----
Closes #1695

### CODE REVIEW

**Comment by elopio** [comment]
Integration test failed with "ConnectionResetError: [Errno 104] Connection reset by peer" Retrying.

File: snapcraft/internal/meta/_desktop.py [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+
+from snapcraft.internal import errors
+
+
+logger = logging.getLogger(__name__)
+
+
+class DesktopFile:
+
+    def __init__(self, *, name, filename, snap_name, prime_dir):
```
Comment (sergiusens) [inline snapcraft/internal/meta/_desktop.py]: How about some type hints here?

File: snapcraft/internal/meta/_errors.py [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+
+class CommandError(Exception):
```
Comment (sergiusens) [inline snapcraft/internal/meta/_errors.py]: Interesting, we missed the opportunity to make this inherit from `SnapcraftError`

**Comment by sergiusens** [comment]
To be clear, this is a plain refactor, right?

**Pull request merged**